File: backend/app/schemas/curriculum.py
"""Curriculum schemas - 教材改版 / 章节"""
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import BaseResponse


class CurriculumChangeItem(BaseModel):
    """单个改版项 (存为 JSON 数组里的元素)"""
    chapter: str
    type: str = Field(description="new / adjusted / removed / renamed")
    summary: str
    key_points: list[str] = Field(default_factory=list)


class CurriculumChangeResponse(BaseResponse):
    """教材改版详情"""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    grade: int
    subject: str
    version: str
    semester: str  # e.g. "2024-fall"

    changes: list[CurriculumChangeItem]

    source: str
    source_url: Optional[str] = None
    verified: bool

    created_at: datetime
    updated_at: datetime


class CurriculumChangeListItem(BaseModel):
    """教材改版列表项"""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    grade: int
    subject: str
    version: str
    semester: str
    change_count: int = 0
    verified: bool


class VideoItem(BaseModel):
    """微课条目 (指向 basic.sh.smartedu.cn 公开微课)"""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    episode: str
    teacher: Optional[str] = None
    school: Optional[str] = None
    duration: Optional[str] = None
    description: Optional[str] = None

    # 链接: 优先 direct_url (人工核对) > search_url (关键词搜索) > chapter_listing_url (整章)
    direct_url: Optional[str] = None
    search_url: Optional[str] = None
    chapter_listing_url: Optional[str] = None

    importance: int = 3
    week_index: Optional[int] = None


class ChapterWeekly(BaseModel):
    """本周要学的章节"""
    chapter: str
    name: str
    week_index: int = Field(description="第几周 (1-based)")
    knowledge_point_count: int = 0
    knowledge_point_codes: list[str] = Field(default_factory=list)
    # 推荐微课 (按 importance 倒序, 取前 5)
    videos: list[VideoItem] = Field(default_factory=list)
    # 兼容旧字段: 取第一个 video 的 direct_url 或 search_url
    video_url: Optional[str] = None
    note: Optional[str] = None


class WeeklyChaptersResponse(BaseModel):
    """本周章节响应"""
    grade: int
    subject: str
    version: str
    semester: str
    week_index: int
    chapters: list[ChapterWeekly]
    is_new_textbook: bool = Field(default=False, description="是否使用新教材")
    change_notice: Optional[str] = None
    # 整周所有推荐微课 (跨章节)
    weekly_videos: list[VideoItem] = Field(default_factory=list)


class ChapterListResponse(BaseModel):
    """某学期全部章节"""
    grade: int
    subject: str
    version: str
    semester: str
    chapters: list[str]
    new_textbook_chapters: list[str] = Field(default_factory=list, description="本学期新教材改动的章节")


__all__ = [
    "CurriculumChangeItem",
    "CurriculumChangeResponse",
    "CurriculumChangeListItem",
    "ChapterWeekly",
    "WeeklyChaptersResponse",
    "ChapterListResponse",
]

File: web/src/lib/curriculum.ts
/**
 * 教材 API
 */
import { api } from './api';
import type { SuccessResponse } from './types';

export interface CurriculumChangeItem {
  chapter: string;
  type: 'new' | 'adjusted' | 'removed' | 'renamed';
  summary: string;
  key_points: string[];
}

export interface CurriculumChange {
  id: number;
  grade: number;
  subject: string;
  version: string;
  semester: string;
  changes: CurriculumChangeItem[];
  source: string;
  source_url: string | null;
  verified: boolean;
  created_at: string;
  updated_at: string;
}

export interface CurriculumChangeListItem {
  id: number;
  grade: number;
  subject: string;
  version: string;
  semester: string;
  change_count: number;
  verified: boolean;
}

export interface VideoItem {
  id: string;
  episode: string;
  teacher: string | null;
  school: string | null;
  duration: string | null;
  description: string | null;
  direct_url: string | null;
  search_url: string | null;
  chapter_listing_url: string | null;
  importance: number;
  week_index: number | null;
}

export interface ChapterWeekly {
  chapter: string;
  name: string;
  week_index: number;
  knowledge_point_count: number;
  knowledge_point_codes: string[];
  videos: VideoItem[];
  video_url: string | null;
  note: string | null;
}

export interface WeeklyChaptersData {
  grade: number;
  subject: string;
  version: string;
  semester: string;
  week_index: number;
  chapters: ChapterWeekly[];
  is_new_textbook: boolean;
  change_notice: string | null;
  weekly_videos: VideoItem[];
}

export interface ChapterListData {
  grade: number;
  subject: string;
  version: string;
  semester: string;
  chapters: string[];
  new_textbook_chapters: string[];
}

export const curriculumApi = {
  async listChanges(opts: {
    grade?: number;
    subject?: string;
    version?: string;
    verifiedOnly?: boolean;
  } = {}): Promise<CurriculumChangeListItem[]> {
    const resp = await api.get<SuccessResponse<CurriculumChangeListItem[]>>(
      '/curriculum/changes',
      { params: opts }
    );
    return resp.data.data;
  },

  async getChange(id: number): Promise<CurriculumChange> {
    const resp = await api.get<SuccessResponse<CurriculumChange>>(
      `/curriculum/changes/${id}`
    );
    return resp.data.data;
  },

  async listChapters(opts: {
    grade: number;
    subject: string;
    version: string;
    semester?: string;
  }): Promise<ChapterListData> {
    const resp = await api.get<SuccessResponse<ChapterListData>>(
      '/curriculum/chapters',
      { params: opts }
    );
    return resp.data.data;
  },

  async weekly(opts: {
    grade: number;
    subject: string;
    version: string;
    semester: string;
  }): Promise<WeeklyChaptersData> {
    const resp = await api.get<SuccessResponse<WeeklyChaptersData>>(
      '/curriculum/weekly',
      { params: opts }
    );
    return resp.data.data;
  },

  /**
   * 真实微课 (从 basic.sh.smartedu.cn API 抓的, 含 resource_id + courseId)
   * 用于 hero "本周推荐", 跳转到 airClassroomTaskDetail 直达视频播放器
   */
  async realVideos(opts: {
    grade: number;
    subject: string;
    version: string;
    semester: string;
    limit?: number;
  }): Promise<RealVideoItem[]> {
    const resp = await api.get<SuccessResponse<{ items: RealVideoItem[]; total: number }>>(
      '/smartedu-videos',
      { params: { ...opts, limit: opts.limit ?? 10 } }
    );
    return resp.data.data.items;
  },
};

/** 真实微课 item (来自 basic.sh.smartedu.cn) */
export interface RealVideoItem {
  id: string;
  grade: number;
  subject: string;
  subject_label: string;
  version: string;
  version_label: string;
  semester: string;
  term_label: string;
  resource_id: string;
  subject_id: string;
  title: string;
  school: string | null;
  teacher: string | null;
  upload_file_name: string | null;
  release_time: string | null;
  video_url: string | null;
  direct_url: string;
  importance: number;
  verified: boolean;
}
